File: linked-list/mid_difficulty/2299-merge-nodes-in-between-zeros/merge-nodes-in-between-zeros.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:

    ListNode* extractNode(ListNode* head){
        if(head == nullptr){
            return nullptr;
        }

        ListNode* crr = head->next;
        int sum = 0;
        while(crr != nullptr && crr->val != 0){
            sum+= crr->val;
            crr = crr->next;
        }

        return new ListNode(sum);
    }

    ListNode* mergeNodes(ListNode* head) {
        if(head == nullptr){
            return nullptr;
        }

        ListNode* crr = head->next;
        ListNode* result = extractNode(head);
        ListNode* crrResult = result;
        while(crr != nullptr){
            if(crr->val == 0 && crr->next != nullptr){
                crrResult->next = extractNode(crr);
                crrResult = crrResult->next;
            }

            crr = crr->next;
        }

        return result;
    }
};

File: 2299-merge-nodes-in-between-zeros/merge-nodes-in-between-zeros.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeNodes(ListNode head) {
        ListNode crr = head.next;
        int sumNodes = 0;

        ListNode result = new ListNode(0);
        ListNode crrResult = result;
        
        while(crr != null){
            if(crr.val != 0){
                sumNodes += crr.val;
            }else{
                crrResult.next = new ListNode(sumNodes);
                crrResult = crrResult.next;
                sumNodes = 0;
            }

            crr = crr.next;
        }

        return result.next;
    }
}